app/src/main/java/edu/mills/heartoftaiwanese/activity/home/HomeFragment.kt

```diff
@@ -0,0 +1,213 @@
+            )
+        }
+        binding.clearButton.setOnClickListener {
+            if (currentTab == 0) {
+                binding.editTextEng.setText("")
+            }
+            if (currentTab == 1) {
+                binding.editTextCh.setText("")
+            }
+            binding.twResult.visibility = View.GONE
+            binding.result.visibility = View.GONE
+        }
+        binding.tabsEnglishChinese.addOnTabSelectedListener(this)
+    }
+
+    private fun hideForSubmit() {
+        binding.twResult.visibility = View.GONE
+        binding.result.visibility = View.GONE
+        binding.progressBarLoading.visibility = View.VISIBLE
+        binding.submitCh.visibility = View.GONE
+        binding.submitEn.visibility = View.GONE
+        binding.clearButton.visibility = View.INVISIBLE
+        hideKeyboard()
+    }
+
+    private fun showAfterSubmit() {
+        binding.progressBarLoading.visibility = View.GONE
+        binding.twResult.visibility = View.VISIBLE
+        binding.result.visibility = View.VISIBLE
+        if (currentTab == 1) {
+            binding.submitCh.visibility = View.VISIBLE
+            binding.editTextCh.visibility = View.VISIBLE
+        }
+        if (currentTab == 0) {
+            binding.submitEn.visibility = View.VISIBLE
+            binding.editTextEng.visibility = View.VISIBLE
+        }
+        binding.clearButton.visibility = View.VISIBLE
+    }
+
+    override fun onChineseFetched(chinese: String) {
+        viewModel.fetchTaiwanese(chinese)
+    }
+
+    override fun onTaiwaneseFetched(taiwanese: String) {
+        binding.result.text = taiwanese
+        activity?.runOnUiThread {
+            showAfterSubmit()
+        }
+    }
+
+    override fun onNetworkError(error: WebResultCode) {
+        Log.e("HomeFragment", error.toString())
+        activity?.runOnUiThread {
+            showAfterSubmit()
+            when (error) {
+                WebResultCode.RATE_LIMITED -> Toast.makeText(
+                    activity,
+                    getText(R.string.too_many_requests),
+                    Toast.LENGTH_LONG
+                ).show()
+                WebResultCode.INVALID_NOT_FOUND -> Toast.makeText(
+                    activity,
+                    getText(R.string.errorNotFound),
+                    Toast.LENGTH_LONG
+                ).show()
+                WebResultCode.UNKNOWN_ERROR -> Toast.makeText(
+                    activity,
+                    getText(R.string.error),
+                    Toast.LENGTH_LONG
+                ).show()
+                // Bad state
+                WebResultCode.RESULT_OK -> throw IllegalStateException("Do not call error when there's no error")
+            }
+        }
+    }
+
+    /**
+     * Called when a tab enters the selected state.
+     *
+     * @param tab The tab that was selected
+     */
+    override fun onTabSelected(tab: TabLayout.Tab?) {
+        currentTab = tab?.position
+        if (currentTab == 1) {
```

Comment:
Consider abstracting out this pattern where you do one thing if `currentTab` is 0 and another if 1. Let me know if you'd like to discuss.